Restrict assembly of face values to discontinuous fields

gassmoeller opened the pull request:
This is a workaround to an inefficiency in our current assembly system. We determine whether to assemble face terms by checking the number of face assemblers. These assemblers might only be necessary for a different advection system though (e.g. the discontinuous compositional fields in case of active particles). This slows down assembly unnecessarily by evaluating FeFaceValues in many unnecessary cases in models with mixed discontinuous/continuous advection elements.

I call this a workaround, because in the future we might have other reasons for assembling face terms for the advection assembly than just discontinuous elements, but since there is no case I am currently aware of, this is the easiest solution for now.

bangerth commented:
Do I understand you right that you are only concerned about the case where *some*, but *not all* compositional or temperature fields are discontinuous, but that we call the face assemblers for all?

I worry that the workaround you have is unnecessarily obscure. We can have boundary integrals for physical reasons, e.g., if one wanted to prescribe inflow boundary conditions weakly, rather than through constraints. I can come up with other reasons where you happen to have boundary integrals even for continuous fields. In those cases, one would attach an assembler that, for hard to find reasons, will not be executed. @jaustermann would have been tripped up on this issue just this week :-)

So if my interpretation is correct, why not call these assemblers but in the ones we have, exit right away if the field in question is not discontinuous?

tjhei commented:
> So if my interpretation is correct, why not call these assemblers but in the ones we have, exit right away if the field in question is not discontinuous?

This is what we currently do, but you still evaluate material models, fevalues, etc.

The correct way to do it would be to add a function like ``bool have_interior_face_intregral_for_advection(field)`` to the assembly interface and loop through the assemblers to find out if this loop is needed.

gassmoeller commented:
Yes, I worry about the case where some of the advection fields (but not all) need face assemblers.
Currently the individual assemblers exit if they notice they are not needed for this field, but at this point the damage is already done, because evaluation of FEFaceValues has already happened. Timo is right the correct way is provide a function that answers the question if we need face assemblers on a per field basis. I will look into that.

gassmoeller commented:
I updated the PR with an improved version that also solves #1388. Assembler can now request to get the viscosity, or to get an initialized FEFaceValues on a per field basis (if they do not request FEFaceValues, they are not called at all).
I still think this whole structure needs a rework, but this change should at least circumvent our current problems.

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -612,6 +612,9 @@ namespace aspect
     aspect::Assemblers::AdvectionAssembler<dim> *adv_assembler
       = new aspect::Assemblers::AdvectionAssembler<dim>();
 
+    assemblers->advection_system_assembler_properties.resize(1+parameters.n_compositional_fields);
```
can you explicitly initialize to false by passing that as the second argument?

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -1438,7 +1463,7 @@ namespace aspect
                   scratch.face_mesh_velocity_values);
 
-            if (assemblers->advection_system_assembler_on_face_properties.need_face_material_model_data)
+            if (assemblers->advection_system_assembler_on_face_properties[advection_field_index].need_face_material_model_data)
```
can we set the values to something invalid if this is false, so it will crash if someone forgets to set the flag?

gassmoeller commented:
I addressed the comment by adding an explicit initialization to `false` to the default constructor of Properties (as far as I am aware std::vector used the default constructor to initialize the members in a call to resize?). 

Unfortunately, I do not see how we can initialize the value to something invalid, but I do not think this would be useful as well, because it would imply that assemblers that do not want to add face terms at all, would need to explicitly disable this property. I do not think this is necessary. I would rather argue that cell terms are the *normal* assembler terms, and if an assembler wants to assemble face terms it needs to adjust its properties.

Another change that I made also removes the allocation of face scratch members when they are not needed for the current field. This save another 30% of temperature assembly time (have not tracked down where this comes from, but it looks like the call to the constructor to FESubfaceValues is extremely expensive). The total temperature assembly time for a model with discontinuous compositional fields (advected by particles) is now exactly the same as for a model with continuous compositional fields (down by 50% compared to the master before this PR, for a 3D spherical model; 2D and box might be affected less).

gassmoeller commented:
I should add that my test model only has 96 cells, so for larger models the constructor calls might not be as important (only done once per timestep), but still I think it makes sense to disable them as well.

tjhei commented:
> as far as I am aware std::vector used the default constructor to initialize the members in a call to resize?

Yes.

> I should add that my test model only has 96 cells, so for larger models the constructor calls might not be as important (only done once per timestep), but still I think it makes sense to disable them as well.

I was about to comment on that, yes. I agree.

bangerth commented:
On 02/24/2017 08:40 AM, Timo Heister wrote:
>     as far as I am aware std::vector used the default constructor to
>     initialize the members in a call to resize?
>
> Yes.

Maybe. It may also be that it calls placement-new which for POD types 
such as bool doesn't do anything at all with a memory location. I 
wouldn't rely on that.

tjhei commented:
> Maybe. It may also be that it calls placement-new which for POD types
> such as bool doesn't do anything at all with a memory location. I
> wouldn't rely on that.

But ``Properties`` is not a POD or trivial class because it has a user-defined constructor. That means resize will have to call it. So, initializing all members in the constructor (like Rene did) is the right thing to do.

bangerth commented:
On 02/24/2017 09:24 AM, Timo Heister wrote:
> But |Properties| is not a POD or trivial class because it has a
> user-defined constructor. That means resize will have to call it.
> So, initializing all members in the constructor (like Rene did) is
> the right thing to do.

Correct. I just wanted to emphasize that being explicit about 
initialization is almost always the right way to go :-)

bangerth commented:
@gassmoeller -- Did you change your mind about this and delete the branch on purpose?

gassmoeller commented:
@tjhei merged this 4 days ago, before I deleted the branch. It is in master already.

bangerth commented:
Oh, I see. It didn't close #1388 right away. In any case, thanks!
